File: Objective C/Direct Integration/ThumbnailExample/ThumbnailExample/ViewController.m
//
//  ViewController.m
//  ThumbnailExample
//
//  Copyright © 2020 Ogury Co. All rights reserved.
//

#import "ViewController.h"
#import "BlackListViewController.h"
#import <OguryChoiceManager/OguryChoiceManager.h>
#import <OguryAds/OguryAds.h>


@interface ViewController()<OguryAdsThumbnailAdDelegate>

@property (nonatomic, weak) IBOutlet UILabel *statusLabel;
@property (nonatomic, strong) OguryAdsThumbnailAd *thumbnail;
@property (nonatomic, assign) BOOL isAdLoaded;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    self.statusLabel.text = @"Choice Manager Loading...";
    
    //The setup of Ogury Choice Manager and Ogury Ads is done AppDelegate file.
    [[OguryChoiceManager sharedManager] askWithViewController:self andCompletionBlock:^(NSError * _Nullable error, OguryChoiceManagerAnswer answer) {
        if (!error) {
            switch (answer) {
                case OguryChoiceManagerAnswerNoAnswer:
                    self.statusLabel.text = @"Choice Manager No Answer";
                    break;
                case OguryChoiceManagerAnswerFullApproval: // TCF Option
                    self.statusLabel.text = @"Choice Manager Full Approval";
                    break;
                case OguryChoiceManagerAnswerPartialApproval: // TCF Option
                    self.statusLabel.text = @"Choice Manager Partial Approval";
                    break;
                case OguryChoiceManagerAnswerRefusal: // TCF Option
                    self.statusLabel.text = @"Choice Manager Refusal";
                    break;
                case OguryChoiceManagerAnswerSaleAllowed: // CCPA Option
                    self.statusLabel.text = @"Choice Manager Sale Allowed";
                    break;
                case OguryChoiceManagerAnswerSaleDenied: // CCPA Option
                    self.statusLabel.text = @"Choice Manager Unknown Option";
                    break;
            }
        } else {
            self.statusLabel.text = [NSString stringWithFormat:@"Choice Manager error : %@", error.description];
        }
    }];
    // Do any additional setup after loading the view.
}

- (IBAction)loadAdBtnPressed:(id)sender {
    self.statusLabel.text = @"Loading Ad...";
    self.thumbnail = [[OguryAdsThumbnailAd alloc]initWithAdUnitID:@"ogury_adunit"];
    self.thumbnail.thumbnailAdDelegate = self;
    
    [self.thumbnail setWhitelistBundleIdentifiers:@[@"com.example.bundle",@"com.example.bundle2"]];// Extenal bundle where thumbnail is allowed to show
    [self.thumbnail setBlacklistViewControllers:@[NSStringFromClass([BlackListViewController class])]];// Blacklisted
    
    [self.thumbnail load];
}

- (IBAction)showAdBtnPressed:(id)sender {
    if (self.isAdLoaded == YES) {
        self.statusLabel.text = @"Ad requested to show";
        [self.thumbnail show];
    }
}

#pragma mark - OguryAds Delegate

- (void)oguryAdsThumbnailAdAdLoaded {
    self.statusLabel.text = @"Ad received";
    self.isAdLoaded = YES;
}

- (void)oguryAdsThumbnailAdAdClosed {
    self.statusLabel.text = @"Ad Closed";
    self.isAdLoaded = NO;
}

- (void)oguryAdsThumbnailAdAdDisplayed {
    self.statusLabel.text = @"Ad on screen";
}

- (void)oguryAdsThumbnailAdAdClicked {
    self.statusLabel.text = @"Ad Clicked";
}

- (void)oguryAdsThumbnailAdAdNotAvailable {
    self.statusLabel.text = @"Ad Not Available";
}

- (void)oguryAdsThumbnailAdAdError:(OguryAdsErrorType)errorType {
    self.statusLabel.text = [NSString stringWithFormat:@"Error : %ld", errorType];
    self.isAdLoaded = NO;
}

@end

File: Swift/Direct Integration/ThumbnailExample/ThumbnailExample/ViewController.swift
//
//  ViewController.swift
//  ThumbnailExample
//
//  Copyright © 2020 Ogury Co. All rights reserved.
//

import UIKit
import OguryChoiceManager
import OguryAds

class ViewController: UIViewController {
    
    @IBOutlet weak var statusTextView: UITextView!
    var thumbnail: OguryAdsThumbnailAd?
    var adLoaded: Bool = false
    
    override func viewDidLoad() {
        super.viewDidLoad()
        addNewStatus("Choice Manager Loading...")
        
        //The setup of Ogury Choice Manager is done AppDelegate.swift file.
        OguryChoiceManager.shared().ask(with: self) { (error, answer) in
            if error == nil {
                switch answer {
                case .noAnswer: // TCF Option
                    self.addNewStatus("Choice Manager No Answer")
                case .fullApproval: // TCF Option
                    self.addNewStatus("Choice Manager Full Approval")
                case .partialApproval: // TCF Option
                    self.addNewStatus("Choice Manager Partial Approval")
                case .refusal: // TCF Option
                    self.addNewStatus("Choice Manager Refusal")
                case .saleAllowed: // CCPA Option
                    self.addNewStatus("Choice Manager Sale Allowed")
                case .saleDenied: // CCPA Option
                    self.addNewStatus("Choice Manager Sale Denided")
                default:
                    self.addNewStatus("Choice Manager Unknown Option")
                }
            } else {
                self.addNewStatus("Choice Manager error : \(error.debugDescription)")
            }
        }
    }

    
    @IBAction func loadAdBtnPressed(_ sender: Any) {
        addNewStatus("Loading Ad...")
        
        thumbnail = OguryAdsThumbnailAd.init(adUnitID: "ogury_adunit")
        thumbnail!.thumbnailAdDelegate = self
        
        thumbnail?.setWhitelistBundleIdentifiers(["com.example.bundle","com.example.bundle2"]) // Extenal bundle where thumbnail is allowed to show
        thumbnail?.setBlacklistViewControllers([NSStringFromClass(BlackListViewController.classForCoder())]) // Blacklisted ViewController where thumbnail is not allowed to show
        
        thumbnail?.load(CGSize(width: 200, height: 200))
    }
    
    @IBAction func showAdBtnPressed(_ sender: Any) {
        guard let thumbnailView = thumbnail else {
            return
        }
        if adLoaded == true {
            addNewStatus("Ad requested to show")
            thumbnailView.show(with: OguryRectCorner.bottomRight, margin: OguryOffset(x: 20, y: 20))
        }
    }

    func addNewStatus(_ status: String) {
        DispatchQueue.main.async {
            let textToLog = status + "\n"
            self.statusTextView.textStorage.append(NSAttributedString(string: textToLog))
            let bottom = NSMakeRange(self.statusTextView.text.count - 1, 1)
            self.statusTextView.scrollRangeToVisible(bottom)
        }
    }
    
}

extension ViewController : OguryAdsThumbnailAdDelegate {
    func oguryAdsThumbnailAdAdLoaded() {
        addNewStatus("Ad received")
        self.adLoaded = true
    }
    
    func oguryAdsThumbnailAdAdClosed() {
        addNewStatus("Ad Closed")
        self.adLoaded = false
    }
    
    func oguryAdsThumbnailAdAdClicked() {
        addNewStatus("Ad Clicked")
    }
    
    func oguryAdsThumbnailAdAdDisplayed() {
        addNewStatus("Ad on screen")
    }
    
    func oguryAdsThumbnailAdAdNotAvailable() {
        addNewStatus("Ad Not Available")
    }
    
    //For more informations about error codes please refer to our documentation at https://docs.ogury.co/
    func oguryAdsThumbnailAdAdError(_ errorType: OguryAdsErrorType) {
        addNewStatus("Error: \(errorType.rawValue)")
        self.adLoaded = false
    }
}

class BlackListViewController: UIViewController {
    override func viewDidLoad() {
        super.viewDidLoad();
    }
}
